beacon-chain/blockchain/core.go

```diff
@@ -249,3 +256,255 @@ func (b *BeaconChain) getAttestersProposer(seed common.Hash) ([]int, int, error)
+	if attesterField%2 != 0 {
+		voted = true
+	}
+
+	return voted
+}
+
+// applyRewardAndPenalty applies the appropriate rewards and penalties according to
+// whether the attester has voted or not.
+func (b *BeaconChain) applyRewardAndPenalty(index int, voted bool) error {
+	defer b.lock.Unlock()
+	b.lock.Lock()
+
+	if voted {
+		b.state.CrystallizedState.ActiveValidators[index].Balance += params.AttesterReward
+	} else {
+		// TODO : Change this when penalties are specified for not voting
+		b.state.CrystallizedState.ActiveValidators[index].Balance -= params.AttesterReward
+	}
+
+	return b.persist()
+}
+
+// resetAttesterBitfields resets the attester bitfields in the ActiveState to zero.
+func (b *BeaconChain) resetAttesterBitfields() error {
+
+	length := int(len(b.state.CrystallizedState.ActiveValidators) / 8)
+	if len(b.state.CrystallizedState.ActiveValidators)%8 != 0 {
+		length += 1
+	}
+
+	defer b.lock.Unlock()
+	b.lock.Lock()
+
+	newbitfields := make([]byte, length)
+	b.state.ActiveState.AttesterBitfields = newbitfields
+
+	return b.persist()
+}
+
+// resetTotalDeposit clears and resets the total attester deposit to zero.
+func (b *BeaconChain) resetTotalDeposit() error {
+	defer b.lock.Unlock()
+	b.lock.Lock()
+	b.state.ActiveState.TotalAttesterDeposits = 0
+
+	return b.persist()
+}
+
+// setJustifiedEpoch sets the justified epoch during an epoch transition.
+func (b *BeaconChain) setJustifiedEpoch() error {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	justifiedEpoch := b.state.CrystallizedState.LastJustifiedEpoch
+	b.state.CrystallizedState.LastJustifiedEpoch = b.state.CrystallizedState.CurrentEpoch
+
+	if b.state.CrystallizedState.CurrentEpoch == (justifiedEpoch + 1) {
+		b.state.CrystallizedState.LastFinalizedEpoch = justifiedEpoch
+	}
+
+	return b.persist()
+}
+
+// setRewardsAndPenalties checks if the attester has voted and then applies the
+// rewards and penalties for them.
+func (b *BeaconChain) setRewardsAndPenalties(index int) error {
+	bitfields := b.state.ActiveState.AttesterBitfields
+	attesterBlock := (index + 1) / 8
+	attesterFieldIndex := (index + 1) % 8
+	if attesterFieldIndex == 0 {
+		attesterFieldIndex = 8
+	}
+
+	if len(bitfields) < attesterBlock {
+		return errors.New("attester index does not exist")
+	}
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	voted := hasVoted(bitfields, attesterBlock, attesterFieldIndex)
+	if err := b.applyRewardAndPenalty(index, voted); err != nil {
+		return fmt.Errorf("unable to apply rewards and penalties: %v", err)
+	}
+
+	return nil
+}
+
+// Slashing Condtions
+// TODO: Implement all the conditions when the spec is updated
+func (b *BeaconChain) heightEquivocationCondition(validatorIndex int) error {
+	testbool := false
+	testSlash := uint64(1)
+	if !testbool {
+		if err := b.slashStake(validatorIndex, testSlash); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (b *BeaconChain) ffgSurroundCondition(validatorIndex int) error {
```

Comment:
Can you please add 1-2 lines function comments on all these conditions, kind of hard to tell what's going on?